Add troubleshooting of rancher agent-tls-mode (#764)
harvester/docs: docs/troubleshooting/rancher.md

    The output of the script indicates the location of the generated tarball.

For more information, see [The Rancher v2.x Linux log collector script](https://www.suse.com/support/kb/doc/?id=000020191).

> ## Harvester is Pending Importing to Rancher Manager
> 
> After setting the `cluster-registration-url` on Harvester, a deployment `cattle-system/cattle-cluster-agent` is created to import the Harvester cluster to `Rancher Manager`.
> 
> There are some known issues.

Reviewer: ```suggestion
## Importing of Harvester Clusters into Rancher

After the `cluster-registration-url` is set on Harvester, a deployment named `cattle-system/cattle-cluster-agent` is created for importing of the Harvester cluster into Rancher.
```

Reviewer: The level-2 heading (##) is supposed to describe an operation or feature, not the actual issue (for example, "Guest Cluster Log Collection").

This file is part of the "Troubleshooting" section, so we don't need text such as "There are some known issues.".

> ### Pending Due to `unable to read CA file`
> 
> When following logs are observed from the pod `cattle-cluster-agent-*`, Harvester can't be imported to `Rancher Manager`.

Reviewer: ```suggestion
### Import Pending Due to `unable to read CA file` Error

The following error messages in the `cattle-cluster-agent-*` pod logs indicate that the Harvester cluster cannot be imported into Rancher.
```

```
2025-02-13T17:25:22.520593546Z time="2025-02-13T17:25:22Z" level=info msg="Rancher agent version v2.10.2 is starting"
2025-02-13T17:25:22.529886868Z time="2025-02-13T17:25:22Z" level=error msg="unable to read CA file from /etc/kubernetes/ssl/certs/serverca: open /etc/kubernetes/ssl/certs/serverca: no such file or directory"
2025-02-13T17:25:22.529924542Z time="2025-02-13T17:25:22Z" level=error msg="Strict CA verification is enabled but encountered error finding root CA"
```

> ### Root Cause
> 
> `Rancher Manager` added a new setting, `agent-tls-mode`, which allows users to specify if agents will use strict certificate verification when connecting to Rancher. Can be set to strict (which requires the agent to verify the cert using only the Certificate Authority in the cacerts setting) or system-store (which allows the agent to verify the cert using any Certificate Authority in the operating system's trust store).
> 
> This setting will default to strict on new installs of 2.9.0+, but will be system-store on new installs of 2.8 or upgrades to 2.9 from 2.8.
> 
> Refer [Release Node](https://github.com/rancher/rancher/issues/45628#issuecomment-2246152604)

Reviewer: ```suggestion
#### Root Cause

Rancher's `agent-tls-mode` setting controls how Rancher's agents (`cluster-agent`, `fleet-agent`, and `system-agent`) validate Rancher's certificate when establishing a connection. You can set either of the following values:

- `strict`: Rancher's agents only trust certificates generated by the Certificate Authority specified in the cacerts setting.
- `system-store`: Rancher's agents trust any certificate generated by a public Certificate Authority specified in the operating system's trust store.

The default value of this setting depends on the Rancher version and installation type. For more information, see Rancher issue [#45628](https://github.com/rancher/rancher/issues/45628#issuecomment-2246152604).

| Type | Versions | Default Value |
| --- | --- | --- |
| New installation | v2.8 | `system-store` |
| New installation | v2.9 and later | `strict` |
| Upgrade | v2.8 to v2.9 | `system-store` |
```

> ### Solution

Reviewer 2: I think the real solution is that user should add their custom CA cert (e.g., Let's Encrypt or bring-your-own) to the rancher `cacerts` settings, as documented at https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/installation-references/tls-settings#agent-tls-enforcement. 

There is more information on configuration steps at https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/install-upgrade-on-a-kubernetes-cluster#3-choose-your-ssl-configuration.

Changing the settings from `strict` to `system store` is less secure because:

> This can be a security risk, since any certificate generated by these external authorities, which are outside the user's control, are considered valid in this state.

I think we should present both options with the settings change being the last resort.

Author: Thanks @ihcsim , the insight is very valuable, I will update the doc with this additional option.

> ### Solution
> 
> 1. Login to the `Rancher-Manager` UI, then navigate to **Global Setttings** > **Setttings**.
> 1. Find and select **agent-tls-mode**, and then select **⋮** > **Edit Setting** to access the configuration options.
> 1. Set the **Value** to `System Store`.
> 
> ![](/img/v1.4/troubleshooting/rancher-global-setting-agent-tls-mode.png)
> 
> Related issues:
>   - [[BUG] Rancher integration w/ Harv v1.4.0 and Rancher v2.9.2 failing](https://github.com/harvester/harvester/issues/7105)
>   - [[QUESTION] Upgrading rancher-vcluster addon?]https://github.com/harvester/harvester/issues/7284
>   - [(Rancher Manager) add new agent-tls-mode setting](https://github.com/rancher/rancher/issues/45628)

Reviewer: ```suggestion
#### Solution

1. Log in to the Rancher UI.

1. Go to **Global Settings > Settings**.

1. Select **agent-tls-mode**, and then select **⋮ > Edit Setting** to access the configuration options.

  ![](/img/v1.4/troubleshooting/rancher-global-setting-agent-tls-mode.png)

1. Set **Value** to **System Store**.

1. Click **Save**.

Related issues:

- Harvester: [#7105](https://github.com/harvester/harvester/issues/7105) and [#7284]https://github.com/harvester/harvester/issues/7284
- Rancher: [#45628](https://github.com/rancher/rancher/issues/45628)
```

    The output of the script indicates the location of the generated tarball.

For more information, see [The Rancher v2.x Linux log collector script](https://www.suse.com/support/kb/doc/?id=000020191).

## Importing of Harvester Clusters into Rancher

After the `cluster-registration-url` is set on Harvester, a deployment named `cattle-system/cattle-cluster-agent` is created for importing of the Harvester cluster into Rancher.

### Import Pending Due to `unable to read CA file` Error

The following error messages in the `cattle-cluster-agent-*` pod logs indicate that the Harvester cluster cannot be imported into Rancher.

```
2025-02-13T17:25:22.520593546Z time="2025-02-13T17:25:22Z" level=info msg="Rancher agent version v2.10.2 is starting"
2025-02-13T17:25:22.529886868Z time="2025-02-13T17:25:22Z" level=error msg="unable to read CA file from /etc/kubernetes/ssl/certs/serverca: open /etc/kubernetes/ssl/certs/serverca: no such file or directory"
2025-02-13T17:25:22.529924542Z time="2025-02-13T17:25:22Z" level=error msg="Strict CA verification is enabled but encountered error finding root CA"
```

> The root cause is that the `agent-tls-mode` is not effectively configured.

Reviewer 2: ```suggestion
The root cause is that the `agent-tls-mode` setting is not correctly configured. The default value of this setting depends on the Rancher version and installation type. For more information, see Rancher issue [#45628](https://github.com/rancher/rancher/issues/45628#issuecomment-2246152604).
| Type | Versions | Default Value |
| --- | --- | --- |
| New installation | v2.8 | `system-store` |
| New installation | v2.9 and later | `strict` |
| Upgrade | v2.8 to v2.9 | `system-store` |
```

Author: As this is a breaking change from Rancher even though it is for better security. When then user upgraded to new and was blocked, the `not effectively configured` feels better than `not correctly configured`.

Reviewer: "correctly" is neutral and straightforward. "effectively" seems to imply both correctness and optimization. If we use "effectively", we have to describe exactly how users can effectively configure the setting.

When documenting settings, we usually start with its purpose or use cases. That said, I suggest keeping the current structure. If users already know about the setting, they can just skip the description and go straight to the table.

    The `strict` option enables a higher level of security, it requires Rancher to have access to the CA which generated the certificate visible to the agents. In the case of certain certificate configurations (notably, external certificates), this is not automatic, and **extra configuration is needed**. See the [installation guide](https://ranchermanager.docs.rancher.com/getting-started/installation-and-upgrade/install-upgrade-on-a-kubernetes-cluster#3-choose-your-ssl-configuration) for more information on which scenarios require extra configuration.

- `system-store`: Rancher's agents trust any certificate generated by a public Certificate Authority specified in the operating system's trust store. Use this setting if your setup uses an external trust authority and you don't have ownership over the Certificate Authority.

    :::important

    Using the `system-store` setting implies that the agent trusts all external authorities found in the operating system's trust store including those outside of the user's control.

    :::

> The default value of this setting depends on the Rancher version and installation type. For more information, see Rancher issue [#45628](https://github.com/rancher/rancher/issues/45628#issuecomment-2246152604).

Reviewer 2: Maybe move this to line 43? See suggestion above.

| Type | Versions | Default Value |
| --- | --- | --- |
| New installation | v2.8 | `system-store` |
| New installation | v2.9 and later | `strict` |
| Upgrade | v2.8 to v2.9 | `system-store` |

> Follow the steps below to configure the different TLS setting options.

Reviewer 2: Maybe indent this to make it part of the `system-store` setting bullet section, after moving line 59 to line 43:

```suggestion
   Follow the steps below to change the `agent-tls-mode` setting to `system-store`:
```

Author: This config fits both `system-store` and `strict`.

The above context has pointed the differences between those two modes and listed the suggestions. To this step, it is the user's choice to set either of the mode.

e.g. when the user can't config `strict` effectively right now, he can fallback to `system-store`, and when investigated everything clearly, then re-configure.

Author: @jillian-maroket It is up to your decision about how the final PR is organized, please take a look when you are free, thanks.

Author: the continuous update of code and doc, and out of sync here and there, make me not so confident that user can reach the target when they just `follow the instructions` once.  back-and-force is likely in the normal path :)

Reviewer: This is a straightforward configuration procedure that applies to both previously described options. Perhaps we can keep the current structure and then add/move details later.
